forge-ai/src/main/java/forge/ai/SpellAbilityAi.java

```diff
@@ -410,4 +413,61 @@ public CounterType chooseCounterType(List<CounterType> options, SpellAbility sa,
     public boolean chooseBinary(BinaryChoiceType kindOfChoice, SpellAbility sa, Map<String, Object> params) {
         return MyRandom.getRandom().nextBoolean();
     }
+
+    public List<OptionalCostValue> chooseOptionalCosts(SpellAbility chosen, Player player, List<OptionalCostValue> optionalCostValues) {
+        List<OptionalCostValue> chosenOptCosts = Lists.newArrayList();
+        Cost costSoFar = chosen.getPayCosts().copy();
+
+        outer:
+        for (OptionalCostValue opt : optionalCostValues) {
+            // Choose the optional cost if it can be paid (to be improved later, check for playability and other conditions perhaps)
+            Cost fullCost = opt.getCost().copy().add(costSoFar);
+            SpellAbility fullCostSa = chosen.copyWithDefinedCost(fullCost);
+
+            // Playability check for Kicker
+            if (opt.getType() == OptionalCost.Kicker1 || opt.getType() == OptionalCost.Kicker2) {
+                SpellAbility kickedSaCopy = fullCostSa.copy();
+                kickedSaCopy.addOptionalCost(opt.getType());
+                Card copy = CardCopyService.getLKICopy(chosen.getHostCard());
+                copy.setCastSA(kickedSaCopy);
+                if (ComputerUtilCard.checkNeedsToPlayReqs(copy, kickedSaCopy) != AiPlayDecision.WillPlay) {
+                    continue; // don't choose kickers we don't want to play
+                }
+            }
+
+            if (ComputerUtilCost.canPayCost(fullCostSa, player, false)) {
+                // check for additional Cost breaking targets AI had already chosen
+                boolean targeting = false;
+                SpellAbility sub = fullCostSa;
+                while (sub != null) {
```

Comment:
doesn't make sense imo
AI would choose optional cost before API sets targets anyway so this might just stop it from playing them at all

Reply:
normally yes, targets would be set after optional Cost, but not for AI
AI would set targets before the spell is added to the Stack (and optional Costs are chosen)

tested with `Into the Flood Maw`:
- without gift, bounce creature
- with gift, bounce nonland

at this point in the code, right now, ChangeZoneAi didn't know about Gift
which means, it would set the Target on the "bounce creature" SA

if now AI would choose to add the Optional Cost, both SA become invalid

my next work needs to be teach AI in ChangeZoneAi about Gift and what that means for targets.
So that AI can use it to bounce nonlands (that are better than a fish token)